Docker "tip"

qdm12 opened the pull request:
This is a mix of different pull requests (opened on https://github.com/zylon-ai/private-gpt/pulls and on https://github.com/qdm12/private-gpt/pulls) regarding Docker, and is the tip of the changes.

You can review each commit from this pull request to see what was changed.

Directions to build and use the Docker image from this branch:

1. Install `git` and `docker` (or `podman` etc.)
2. Build the image from this branch:

    ```sh
    docker build -t privategpt --build-arg POETRY_EXTRAS="ui vector-stores-qdrant embeddings-huggingface llms-ollama embeddings-ollama" --build-arg UID=100 --build-arg GID=65534 https://github.com/qdm12/private-gpt.git#docker-tip
    ```

    You can add and remove extras to `POETRY_EXTRAS`, for example for Vllm, add ` llms-openailike`.
3. Create directories to bind mount with the right permissions:

    ```sh
    mkdir -p local_data cache/models cache/tiktoken
    chown -R 100:65534 local_data
    chown -R 100:65534 cache
    ```

    Note you can set the build arguments UID and GID to change these values, OR add a `user: 100:65534` to your docker-compose.yml to run as a specific uid and gid.
4. You can run it with for example [the updated docker-compose.yml](https://github.com/qdm12/private-gpt/blob/docker-tip/docker-compose.yaml). Obviously you might need to adapt environment variables to match what you want to do/configure. I personally use vllm for the llm backend (`PGPT_MODE=openailike`) and `PGPT_EMBEDDING_MODE=huggingface`.

If you encounter any issue, please [open an issue on this repo](https://github.com/qdm12/private-gpt/issues/new) and I'll address it, and PR it to the base repository. Doing all of this since the zylon-ai repository maintainers look rather slow at reviewing/merging PRs. I'll also do my best to keep this fork updated with the base repository.

qdm12 committed:
`UID` and `GID` build arguments for `worker` user

qdm12 committed:
`POETRY_EXTRAS` build argument with default values

qdm12 committed:
Copy `Makefile` for `make ingest` command

qdm12 committed:
Do NOT copy markdown files
I doubt anyone reads a markdown file within a Docker image

qdm12 committed:
Fix PYTHONPATH value

qdm12 committed:
Set home directory to `/home/worker` when creating user

qdm12 committed:
Combine `ENV` instructions together

qdm12 committed:
Define environment variables with their defaults
- For documentation purposes
- Reflect defaults set in settings-docker.yml

qdm12 committed:
`PGPT_EMBEDDING_MODE` to define embedding mode

qdm12 committed:
chore(docker): merge local and external dockerfiles
- llama-cpp dependencies install does not occupy space in final image
- llama-cpp dependencies install takes about 10 seconds, worth the trade off
- Update docker.yml workflow to use the standard `Dockerfile`

qdm12 committed:
Remove ineffective `python3 -m pipx ensurepath`

qdm12 committed:
Use `&&` instead of `;` to chain commands to detect failure better

qdm12 committed:
Move llama-cpp dependencies install to intermediary dependencies layer

qdm12 committed:
Define openai section with variables
- Add openai section in settings-docker.yaml
- Define variable `PGPT_OPENAI_API_BASE`
- Define variable `PGPT_OPENAI_API_KEY`
- Define variable `PGPT_OPENAI_MODEL`

qdm12 committed:
Change default `PGPT_EMBEDDING_MODE` from `sagemaker` to `mock`

qdm12 committed:
Add `--no-root` flag to poetry install commands

qdm12 committed:
Specify image name `privategpt`

qdm12 committed:
Add `embeddings-huggingface` to POETRY_EXTRAS

qdm12 committed:
Add Nvidia GPU support definitions

qdm12 committed:
Remove environment variables set to their defaults

qdm12 committed:
Add bind mounts for caches

qdm12 committed:
Set PGPT_PROFILES to docker

qdm12 committed:
Copy variables from Dockerfile and set them to their defaults

qdm12 committed:
Expose port of Ollama

qdm12 committed:
Change commented context to fork for now

levydanqc commented:
Hi! Running this, I get the error: 
`PermissionError: [Errno 13] Permission denied: '/home/worker/app/models/cache/models--mistralai--Mistral-7B-Instruct-v0.2'`
Do you know why please ?

Here were the steps:
```
git clone git@github.com:qdm12/private-gpt.git && cd private-gpt
git checkout docker-tip
docker build -t privategpt --build-arg POETRY_EXTRAS="ui llms-ollama embeddings-ollama vector-stores-qdrant" https://github.com/qdm12/private-gpt.git#docker-tip
vi docker-compose.yaml -> changes: POETRY_EXTRAS=ui llms-ollama embeddings-ollama vector-stores-qdrant
docker compose up -d
```
The Ollama docker is running fine, but the privategpt docker exited with failure (the error above).

qdm12 commented:
I've added missing steps, I think the commands in step 3 should do to fix the ownership of the data and models directory.

levydanqc commented:
The error is still happening, but I might have find something else in the logs! Before this error, the Ollama docker yield : `ollama-1  | Error: mkdir /root/.ollama/models/manifests: permission denied`
It probably has something to do with this, no ?

Ps. I don't know if this error was already there or if step 3 created it...

qdm12 commented:
> The error is still happening, but I might have find something else in the logs! Before this error, the Ollama docker yield

I actually don't use the ollama container (I use vllm), so I don't know. That error is irrelevant to step 3 although similar. As a quick fix, you can try `chmod -R 777 models` to allow all users access to the models directory, that should fix the permission issue.

levydanqc commented:
Unfortunately, I tried giving all permissions to the folders inside the main folder, but the error is still there.
I don't know if this: `PermissionError: [Errno 13] Permission denied: '/home/worker/app/models/cache/models--mistralai--Mistral-7B-Instruct-v0.2'` means that it doesn't have permissions to access the folder or the file and the docker exit on the error so I can change the permissions.

qdm12 commented:
Also I fixed step 3, the directory names needed for privategpt are cache and local_data, not data +models:

```
mkdir -p local_data cache/models cache/tiktoken
chown -R 100:65534 local_data
chown -R 100:65534 cache
```

My bad!

Alternatively you can set `user: "root"` in docker-compose.yml although that's not super great security wise, it's better to run without root.

Sprooty commented:
@qdm12,

To make your steps work out of the box, can i suggest two changes,

1) change your default command to
`docker build -t privategpt --build-arg POETRY_EXTRAS="ui vector-stores-qdrant embeddings-huggingface llms-ollama embeddings-ollama" https://github.com/qdm12/private-gpt.git#docker-tip`

2) use $pwd for the local_data and cache folder?

qdm12 commented:
@Sprooty 1. done, but 2. is not needed since docker-compose understands `./models` as `$(pwd)/models`

bruno-viva commented:
Thanks for this! I had to update this in the docker-compose: `PGPT_OLLAMA_API_BASE: http://host.docker.internal:11434`

But now, with the UI mode and the LLM Chat, I get this error with an empty response after submitting a sample:

```
private-gpt-private-gpt-1  | 00:16:41.049 [INFO    ]            uvicorn.access - 172.19.0.1:60168 - "POST /run/predict HTTP/1.1" 200
private-gpt-private-gpt-1  | 00:16:41.050 [INFO    ]            uvicorn.access - 172.19.0.1:60152 - "POST /run/predict HTTP/1.1" 200
private-gpt-private-gpt-1  | 00:16:41.053 [INFO    ]            uvicorn.access - 172.19.0.1:60152 - "POST /run/predict HTTP/1.1" 200
private-gpt-private-gpt-1  | 00:16:41.056 [INFO    ]            uvicorn.access - 172.19.0.1:60152 - "POST /queue/join HTTP/1.1" 200
private-gpt-private-gpt-1  | 00:16:41.059 [INFO    ]            uvicorn.access - 172.19.0.1:60152 - "GET /queue/data?session_hash=h60wow4ofi8 HTTP/1.1" 200
private-gpt-private-gpt-1  | 00:16:41.425 [WARNING ] llama_index.core.chat_engine.types - Encountered exception writing response to history: [Errno -2] Name or service not known
private-gpt-private-gpt-1  | 00:16:41.467 [INFO    ]            uvicorn.access - 172.19.0.1:60152 - "POST /run/predict HTTP/1.1" 200
``` 

Any ideas on what am I missing? Thanks in advance

(attaching image): 
![image](https://github.com/qdm12/private-gpt/assets/56004875/5ecd7676-7825-42ca-8933-b89d2cdcb901)

qdm12 commented:
> [Errno -2] Name or service not known

This is because `host.docker.internal` does not resolve from within the privategpt container. This is a bit out of scope, can you create an issue on the repository to discuss this further? I'll mark these 2 comments as off topic.

qdm12 commented:
Branch rebased on top of https://github.com/zylon-ai/private-gpt/commit/9d0d614706581a8bfa57db45f62f84ab23d26f15

qdm12 committed:
feat(docker): specify RAG settings with environment variables
- PGPT_RAG_SIMILARITY_TOP_K
- PGPT_RAG_RERANK_ENABLED
- PGPT_RAG_RERANK_MODEL
- PGPT_RAG_RERANK_TOP_N
